third_party/xla_client/metrics_analysis.cc

```diff
@@ -0,0 +1,204 @@
+#include "tensorflow/compiler/xla/xla_client/metrics_analysis.h"
+
+#include "absl/types/variant.h"
+#include "tensorflow/compiler/xla/xla_client/computation_client.h"
+#include "tensorflow/compiler/xla/xla_client/metrics.h"
+#include "tensorflow/compiler/xla/xla_client/tf_logging.h"
+#include "tensorflow/compiler/xla/xla_client/types.h"
+
+namespace xla {
+namespace metrics {
+
+namespace {
+
+float DivToFloat(std::ldiv_t val, std::int64_t denominator) {
+  return val.quot + (float) val.rem / denominator;
+}
+
+class MetricFrequency : public Analyzer {
+ public:
+  MetricFrequency(std::string metric_name, float frequency_threshold,
+                  long warmup_steps = 0)
+      : metric_name_(metric_name),
+        frequency_threshold_(frequency_threshold),
+        warmup_steps_(warmup_steps) {}
+
+  Analysis Run() override {
+    CounterData* step = GetCounter("MarkStep");
+    MetricData* metric = GetMetric(metric_name_);
+    if (!step || !metric) {
+      return {Analysis::Symptom::kNormal};
+    }
+    size_t metric_count = metric->TotalSamples();
+    xla::int64 step_count = step->Value();
+    if (step_count <= warmup_steps_) {
+      return {Analysis::Symptom::kNormal};
+    }
+    auto res = std::div(metric_count, step_count);
+    if (DivToFloat(res, step_count) > frequency_threshold_) {
+      return {
+          Analysis::Symptom::kMetricTooFrequent,
+          absl::StrFormat("%s too frequent: %zu counts during %zu steps",
+                          metric_name_, metric_count, step_count),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::string metric_name_;
+  float frequency_threshold_;
+  long warmup_steps_;
+};
+
+class MetricTime : public Analyzer {
```

Comment:
FYI; we do have a humanization bug in our metrics, example: https://github.com/pytorch/xla/issues/2636#issuecomment-732119675 and other comments.